packages/jira-adapter/src/change_validators/workflowsV2/status_mappings.ts

```diff
@@ -0,0 +1,247 @@
+import { WorkflowSchemeItem, isWorkflowSchemeItem, projectHasWorkflowSchemeReference } from '../workflow_scheme_migration'
+import { ISSUE_TYPE_NAME, JIRA_WORKFLOW_TYPE, PROJECT_TYPE } from '../../constants'
+
+const { awu } = collections.asynciterable
+
+type StatusMigration = {
+  oldStatusReference: ReferenceExpression
+  newStatusReference: ReferenceExpression
+}
+
+type StatusMapping = {
+  issueTypeId: ReferenceExpression
+  projectId: ReferenceExpression
+  statusMigrations: StatusMigration[]
+}
+
+const ISSUE_MAPPINGS_SCHEMA = Joi.array().items(Joi.object({
+  issueTypeId: Joi.object().required(),
+  projectId: Joi.object().required(),
+  statusMigrations: Joi.array().items(Joi.object({
+    oldStatusReference: Joi.object().required(),
+    newStatusReference: Joi.object().required(),
+  })).required(),
+})).required()
+
+const isStatusMappingsSchemeGuard = createSchemeGuard<StatusMapping[]>(ISSUE_MAPPINGS_SCHEMA, 'received invalid statusMappings')
+
+const isStatusMappings = (statusMappings: unknown): statusMappings is StatusMapping[] =>
+  isStatusMappingsSchemeGuard(statusMappings)
+  && statusMappings.every(statusMapping =>
+    isReferenceExpression(statusMapping.issueTypeId)
+    && isReferenceExpression(statusMapping.projectId)
+    && statusMapping.statusMigrations.every(
+      statusMigration =>
+        isReferenceExpression(statusMigration.oldStatusReference)
+        && isReferenceExpression(statusMigration.newStatusReference)
+    ))
+
+
+const isSameOldStatusReference = (statusMigration1: StatusMigration, statusMigration2: StatusMigration): boolean =>
+  statusMigration1.oldStatusReference.elemID.isEqual(statusMigration2.oldStatusReference.elemID)
+
+const isSameStatusMappings = (
+  statusMapping1: StatusMapping,
+  statusMapping2: StatusMapping,
+): boolean =>
+  statusMapping1.issueTypeId.elemID.isEqual(statusMapping2.issueTypeId.elemID)
+    && statusMapping1.projectId.elemID.isEqual(statusMapping2.projectId.elemID)
+    && _.isEmpty(_.differenceWith(
+      statusMapping1.statusMigrations,
+      statusMapping2.statusMigrations,
+      isSameOldStatusReference
+    ))
+
+const getRelevantChanges = (
+  changes: ReadonlyArray<Change<ChangeDataType>>
+): ModificationChange<InstanceElement>[] =>
+  changes
+    .filter(isInstanceChange)
+    .filter(isModificationChange)
+    .filter(change => getChangeData(change).elemID.typeName === JIRA_WORKFLOW_TYPE)
+
+const getRemovedStatuses = (change: ModificationChange<InstanceElement>): ReferenceExpression[] => {
+  const beforeStatuses = (change.data.before.value.statuses ?? [])
+    .map((status: Value) => status.statusReference)
+    .filter(isReferenceExpression)
+  const afterStatuses = (change.data.after.value.statuses ?? [])
+    .map((status: Value) => status.statusReference)
+    .filter(isReferenceExpression)
+  return _.differenceWith(
+    beforeStatuses,
+    afterStatuses,
+    (status1: Value, status2: Value) => status1.elemID.isEqual(status2.elemID)
```

Comment:
nit - status is already `ReferenceExpression`, so no need to specify the type explicitly :
```suggestion
    status => status.elemID.getFullName()
```

Reply:
it is a bit different now, but still typescript dont like it when I remove the type there